contrib/extraman/cprng.8
.\" $MirOS: src/libexec/cprng/cprng.8,v 1.3 2007/09/28 20:04:12 tg Exp $
.\"-
.\" Copyright (c) 2007
.\"	Thorsten Glaser <tg@mirbsd.de>
.\"
.\" Provided that these terms and disclaimer and all copyright notices
.\" are retained or reproduced in an accompanying document, permission
.\" is granted to deal in this work without restriction, including un-
.\" limited rights to use, publicly perform, distribute, sell, modify,
.\" merge, give away, or sublicence.
.\"
.\" This work is provided "AS IS" and WITHOUT WARRANTY of any kind, to
.\" the utmost extent permitted by applicable law, neither express nor
.\" implied; without malicious intent or gross negligence. In no event
.\" may a licensor, author or contributor be held liable for indirect,
.\" direct, other damage, loss, or other issues arising in any way out
.\" of dealing in the work, even if advised of the possibility of such
.\" damage or existence of a defect, except proven that it results out
.\" of said person's immediate fault when using the work as intended.
.\"-
.Dd $Mdocdate: September 28 2007 $
.Dt CPRNG 8
.Os
.Sh NAME
.Nm cprng
.Nd CPU Oscillation Random Number Generator
.Sh SYNOPSIS
.Nm
.Op Fl p
.Nm
.Op Fl p
.Fl r Ns Ar number
.Sh DESCRIPTION
.Nm
generates random numbers by taking the difference between the
cyclic clock of the mainboard and the CPU.
This algorithm is often known as
.Dq truerand
and only works if the CPU and timer interrupt are driven by
separate stroke sources, and works better the more idle
the system is, because otherwise basically the scheduler
is measured.
.Pp
Normal mode of operation is to run in the background and
periodically feed some bytes into the kernel
.Xr random 9
subsystem by writes to
.Pa /dev/urandom .
If you have better entropy sources, like a hardware RNG
on some chipsets, or a combination of much disc, network
and interactive I/O, or if your interrupt and CPU clocks
are not driven by separate ticks, you might want to disable
.Nm
in
.Pa /etc/rc.conf .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl p
Use
.Dv ITIMER_PROF
as timer source.
This one is often more reliably from a different clock source than
.Dv ITIMER_REAL ,
but does not exist on some operating systems.
.It Fl r Ns Ar number
Instead of becoming a daemon and running in the background with low
priority, feeding generated entropy into
.Pa /dev/urandom ,
create
.Ar number
bytes of entropy and write it to standard output while running with
high priority, then exit.
The
.Ar number
argument must be decimal and between 1 and 1024, inclusive.
This is the preferred mode of operation on other operating systems.
Output quality can be measured with MirPorts
.Pa ports/math/ent
and this option.
.El
.Sh SEE ALSO
.Xr arc4random 3 ,
.Xr RC4 3 ,
.Xr random 4 ,
.Xr random 9
.Pp
.Pa http://www.cs.berkeley.edu/~daw/rnd/mab-rand
.Pp
.Pa http://fourmilab.ch/random/
.Sh HISTORY
The
.Nm
daemon first appeared in
.Mx 10 .
.Sh AUTHORS
The
.Nm
utility was written by
.An Thorsten Glaser Aq tg@mirbsd.de .
.Sh CAVEATS
The output produced by the
.Fl r
option is debiased using the von Neumann compressor (whitener) algorithm,
but still not yet suitable for direct use.
It can be fed into the kernel
.Xr random 9
subsystem on
.Mx
and
.Ox ,
but not, for example, on
.Fx
or
.Tn GNU/Linux .
Post-processing with
.Xr RC4 3
and mixing with other entropy sources is highly recommended.
.Pp
Measure output quality with John Walker's
.Nm ent
utility before use on any system.
Take special attention to the Chi-Square values.
.Pp
This utility has been retired and is no longer part of
.Mx .


src/libexec/cprng/cprng.c
/*-
 * Copyright (c) 2007
 *	Thorsten Glaser <tg@mirbsd.de>
 *
 * Provided that these terms and disclaimer and all copyright notices
 * are retained or reproduced in an accompanying document, permission
 * is granted to deal in this work without restriction, including un-
 * limited rights to use, publicly perform, distribute, sell, modify,
 * merge, give away, or sublicence.
 *
 * This work is provided "AS IS" and WITHOUT WARRANTY of any kind, to
 * the utmost extent permitted by applicable law, neither express nor
 * implied; without malicious intent or gross negligence. In no event
 * may a licensor, author or contributor be held liable for indirect,
 * direct, other damage, loss, or other issues arising in any way out
 * of dealing in the work, even if advised of the possibility of such
 * damage or existence of a defect, except proven that it results out
 * of said person's immediate fault when using the work as intended.
 */

/*-
 * This is designed after the “truerand” entropy generator written by
 * Don Mitchell.  The basic idea is to measure the difference between
 * the mainboard clock and the CPU clock then pass the output through
 * whitening.  It's been designed for a SPARCstation 20 (under SunOS)
 * and a Pentium-100 (under BSD/OS), so it should be find with MirOS.
 * This is deliberately written to use as few libc calls as possible.
 *
 * The raw output stream has an entropy of about 7.8 bit/byte (97.5%)
 * if 1024 bytes of raw output are run through ports/math/ent, that's
 * why I chose to let random(9) care about post-processing instead of
 * adding arcfour output compression to this programme.
 *
 * See stathz(9) for more information on the -p flag's timer choice.
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#if defined(__OpenBSD__)
#include <dev/rndioctl.h>
#endif
#include <fcntl.h>
#include <signal.h>
#include <stdbool.h>
#include <stdint.h>
#include <unistd.h>

__RCSID("$MirOS: src/libexec/cprng/cprng.c,v 1.13 2008/12/27 22:15:31 tg Exp $");

#ifndef MAYPROF
#if defined(SIGPROF) && defined(ITIMER_PROF)
#define MAYPROF	1
#else
#define MAYPROF	0
#endif
#endif

static volatile sig_atomic_t glocke;
static useconds_t littlesleep = 2000;
static uint8_t obuf[1024];
static uint64_t intropy = 0;
#if MAYPROF
static bool doprof = false;
#endif

static void laeuten(int);
static int getbits(void);
static void getent(uint8_t *, size_t)
    __attribute__((bounded (string, 1, 2)));

#ifdef __MirBSD__
/* to keep us small */
extern void _exit__(int) __dead;
void exit(int);
int atexit(void (*)(void));
#endif

static void
laeuten(int s __attribute__((unused)))
{
	glocke = 1;
}

static int
getbits(void)
{
	unsigned count = 0;
	struct itimerval itv = {
		.it_interval = {
			.tv_sec = 0,
			.tv_usec = 0
		},
		.it_value = {
			.tv_sec = 0,
			.tv_usec = 16667
		}
	};

	glocke = 0;
	setitimer(
#if MAYPROF
	    doprof ? ITIMER_PROF :
#endif
	    ITIMER_REAL, &itv, NULL);
	while (!glocke)
		++count;

	usleep(littlesleep);
	return (count);
}

static void
getent(uint8_t *buf, size_t n)
{
	int value = 0, bits = 0, t1, t2, tb;

	goto getent_entry;

 getent_loop:
	if ((t1 & 1) != (t2 & 1)) {
		value = (value << 1) | (t1 & 1);
		if (++bits == 8) {
			*buf++ = (uint8_t)value;
			if (--n == 0)
				return;
			bits = 0;
		}
	}
	t1 >>= 1;
	t2 >>= 1;
	if (--tb == 0) {
 getent_entry:
		t1 = getbits();
		t2 = getbits();
		tb = 12;
	}
	goto getent_loop;
}

static const char dmsg[] = "Cannot open random device!\n";
static const char emsg[] = "Syntax error!\n";
static const char fmsg[] = "Cannot daemonise!\n";

int
main(int argc, char *argv[])
{
	size_t num;
	int c = 0;
	unsigned u;
	char *cp;

#if MAYPROF
	signal(SIGPROF, laeuten);
#endif
	signal(SIGALRM, laeuten);
	if (argc == 2 && *(cp = argv[1]) == '-') {
		switch (argv[1][1]) {
		case 'r':
			c = 1;
			break;
		case 'p':
			doprof = true;
			break;
		}
		switch (argv[1][2]) {
		case 'r':
			c = 1;
			break;
		case 'p':
			doprof = true;
			break;
		}
		if (c)
			goto out_bytes;
	} else if (argc > 1) {
		write(2, emsg, sizeof (emsg) - 1);
		return (1);
	}

	switch (fork()) {
	case -1:
 edetach:
		write(2, fmsg, sizeof (fmsg) - 1);
		return (3);
	case 0:
		break;
	default:
		return(0);
	}

	if (setsid() == -1)
		goto edetach;

	if (chdir("/") || (c = open("dev/urandom", O_RDWR)) < 0) {
		write(2, dmsg, sizeof (dmsg) - 1);
		return (2);
	}

	close(0);
	close(1);
	close(2);

 main_loop:
	getent(obuf, 12 + 1);
	write(c, obuf, 12);
	u = 94;
	ioctl(c, RNDADDTOENTCNT, &u);
	usleep(littlesleep);
	if (!intropy)
		read(c, &intropy, sizeof (intropy));
	u = intropy & 0x0F;
	intropy >>= 4;
	u = ((u << 4) | u) ^ obuf[12];
	sleep((u & 0x0F) + 7);
	littlesleep = ((u & 0xF0) << 4) + 42000;
	goto main_loop;

 out_bytes:
	while (*cp < '0' || *cp > '9')
		++cp;
	if ((c = cp[0]) > '0' && c <= '9') {
		num = c - '0';
		if ((c = cp[1]) >= '0' && c <= '9') {
			num = num * 10 + c - '0';
			if ((c = cp[2]) >= '0' && c <= '9') {
				num = num * 10 + c - '0';
				if ((c = cp[3]) >= '0' && c <= '9')
					num = num * 10 + c - '0';
			}
		}
		if (num > sizeof (obuf))
			num = sizeof (obuf);
	} else
		num = 16;

	getent(obuf, num);
	write(1, obuf, num);
	return (0);
}

#ifdef __MirBSD__
void
exit(int status)
{
	close(0);
	close(1);
	close(2);
	_exit__(status);
}

int
atexit(void (*function)(void) __attribute__((unused)))
{
	return (-1);
}
#endif
